src/plots/barPlot.ts

```diff
@@ -468,78 +468,133 @@ export module Plots {
-        let tooWide = secondaryAttrTextSpace + 2 * Bar._LABEL_HORIZONTAL_PADDING > secondaryAttrAvailableSpace;
-        if (measurement.height <= h && measurement.width <= w) {
-          let offset = Math.min((primary - primarySpace) / 2, Bar._LABEL_VERTICAL_PADDING);
-          if (!positive) { offset = offset * -1; }
-          if (this._isVertical) {
-            y += offset;
-          } else {
-            x += offset;
+
+        let xAlignment = "center";
+        let yAlignment = "center";
+        let labelContainerOrigin = {
+          x: attrToProjector["x"](d, i, dataset),
+          y: attrToProjector["y"](d, i, dataset)
+        };
+        let containerWidth = barWidth;
+        let containerHeight = barHeight;
+
+        let labelOrigin = {
+          x: labelContainerOrigin.x,
+          y: labelContainerOrigin.y
+        };
+
+        let showLabelOnBar = true;
+
+        if (this._isVertical) {
+          labelOrigin.x += containerWidth / 2 - measurement.width / 2;
+
+          let barY = attrToProjector["y"](d, i, dataset);
+          let effectiveBarHeight = Utils.Math.min([
+            this.height() - barY,
+            barY + barHeight,
+            barHeight
+          ], 0);
+          let offset = Bar._LABEL_VERTICAL_PADDING;
+          showLabelOnBar = measurement.height + 2 * offset <= effectiveBarHeight;
+
+          if (showLabelOnBar) {
+            if (scaledValue < scaledBaseline) {
+              labelContainerOrigin.y += offset;
+              yAlignment = "top";
+              labelOrigin.y += offset;
+            } else {
+              labelContainerOrigin.y -= offset;
+              yAlignment = "bottom";
+              labelOrigin.y += containerHeight - offset - measurement.height;
+            }
+          } else { // show label off bar
+            containerHeight = barHeight + offset + measurement.height;
+            if (scaledValue <= scaledBaseline) {
+              labelContainerOrigin.y -= offset + measurement.height;
+              yAlignment = "top";
+              labelOrigin.y -= offset + measurement.height;
+            } else {
+              yAlignment = "bottom";
+              labelOrigin.y += barHeight + offset;
+            }
           }
-          let showLabel = true;
-          let labelPosition = {
-            x: x,
-            y: positive ? y : y + h - measurement.height
-          };
-
-          if (this._isVertical) {
-            labelPosition.x = x + w / 2 - measurement.width / 2;
-          } else {
-            labelPosition.y = y + h / 2 - measurement.height / 2;
-            if (!positive) {
-              labelPosition.x = x + w - measurement.width;
+        } else { // horizontal
+          labelOrigin.y += containerHeight / 2 - measurement.height / 2;
+
+          let barX = attrToProjector["x"](d, i, dataset);
+          let effectiveBarWidth = Utils.Math.min([
+            this.width() - barX,
+            barX + barWidth,
+            barWidth
+          ], 0);
```

Comment:
Consider `Math.min` as we know that we have 3 entries here.  Having the default case does not seem useful.